pangeo_forge_esgf/utils.py

```diff
@@ -1,9 +1,37 @@
 from typing import Dict
+from .params import id_templates
+import requests
+import re
 
-def facets_from_iid(iid: str) -> Dict[str, str]:
+
+def facets_from_iid(iid: str, mip: str = None) -> Dict[str, str]:
     """Translates iid string to facet dict according to CMIP6 naming scheme"""
-    iid_name_template = "mip_era.activity_id.institution_id.source_id.experiment_id.variant_label.table_id.variable_id.grid_label.version"
+    if mip is None:
+        # take project id from first iid entry by default
+        mip = iid.split(".")[0]
+    iid_name_template = id_templates[mip]
     facets = {}
     for name, value in zip(iid_name_template.split("."), iid.split(".")):
         facets[name] = value
-    return facets
\ No newline at end of file
+    return facets
+
+
+def get_dataset_id_template(project: str, url: str = None):
+    """Requests the dataset_id string template for an ESGF project"""
+    if url is None:
+        url = "https://esgf-node.llnl.gov/esg-search/search"
+    params = {
+        "project": project,
+        "fields": "*",
+        "limit": 1,
+        "format": "application/solr+json"
+    }
+    r = requests.get(url, params)
+    #print(r.status_code)
+    return r.json()["response"]["docs"][0]["dataset_id_template_"][0]
+
+
+def facets_from_template(template: str):
+    """Parse the (dataset_id) string template into a list of (facet) keys"""
+    regex = r"\((.*?)\)"
+    return re.findall(regex, template)
```

Comment:
I dived a little bit into the [ESGF API documentation](https://esgf.github.io/esg-search/ESGF_Search_RESTful_API.html) and added now these two functions that should be able to retrieve the `dataset_id_template_` (and parse it into facets) for any ESGF project. I don't use them yet but i think this is better than what i did in the `params` module, e.g.,
```python
template = utils.get_dataset_id_template("CMIP6")
# returns '%(mip_era)s.%(activity_drs)s.%(institution_id)s.%(source_id)s.%(experiment_id)s.%(member_id)s.%(table_id)s.%(variable_id)s.%(grid_label)s'
```
which can be parsed into:
```python
utils.facets_from_template(template)
```
```
['mip_era',
 'activity_drs',
 'institution_id',
 'source_id',
 'experiment_id',
 'member_id',
 'table_id',
 'variable_id',
 'grid_label']
```

Reply:
hmmm, here we have `member_id` instead of `variant_label`, that seems not right...

Reply:
Oh this was a long standing issue of confusion for me. I believe it is only relevant for DAMIP, but I will need to double check this.